design/one-pager-function-go-templating.md

```diff
@@ -0,0 +1,171 @@
+# Go Templating Function
+ 
+* Owner: Ezgi Demirel (@ezgidemirel)
+* Reviewers: Crossplane Maintainers
+* Status: Accepted
+
+## Background
+
+With Crossplane 1.14, Composition Functions will be promoted to `v1beta1`, and several 
+composition functions will be introduced as an extension point to be used in pipelines. 
+One of these extensions is the [Go Templating Function], which allows users to render Crossplane
+resources using Go templating capabilities like conditionals and loops. In addition to that, 
+users can use values in the environment configs or other resource fields to create/configure
+Crossplane resources.
+
+## Getting Template Files
+
+With the initial implementation of the Go Templating Function, users will be able
+to provide templates in two different ways:
+- An `inline` source can be used to provide template files as a string in the
+  `input.inline.template` field of the pipeline step.
+- A `fileSystem` source can be used to provide template files with folder path
+  in the `input.fileSystem.dirPath` field of the pipeline step.
+
+Both of these options have their own advantages and disadvantages. While `inline` source
+is faster and easier to use, it is not suitable for large template files and hard to maintain.
+On the other hand, `fileSystem` source requires the files to be present in the local file system,
+which might be possible by mounting config maps with `DeploymentRuntimeConfig` or building another
+container image on top of the existing one with the template files. 
+
+In the next iterations, it is planned to support fetching template files from a remote location like
+GitHub repositories or S3 buckets.
+
+## Rendering Templates and Accessing Resources
+
+To make rendering templates and accessing resource fields easier, function requests
+will be converted to `map[string]interface{}` and passed to the template engine. This will
+allow users to access resource fields in the template files like:
+
+```yaml
+{{- range $i := until ( .observed.composite.resource.spec.count | int ) }}
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: User
+metadata:
+  annotations:
+    gotemplating.fn.crossplane.io/composition-resource-name: test-user-{{ $i }}
+  name: test-user-{{ $i }}
+  labels:
+    testing.upbound.io/example-name: test-user-{{ $i }}
+spec:
+  forProvider: {}
+
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: AccessKey
+metadata:
+  name: sample-access-key-{{ $i }}
+spec:
+  forProvider:
+    userSelector:
+      matchLabels:
+        testing.upbound.io/example-name: test-user-{{ $i }}
+  writeConnectionSecretToRef:
+    name: sample-access-key-secret-{{ $i }}
+    namespace: crossplane-system
+{{- end }}
+
+```
+
+To compare and prepare the desired composed resources map with consistent keys, `function-go-templating` requires a special
+annotation `gotemplating.fn.crossplane.io/composition-resource-name` to be assigned to the input resources. This
```

Comment:
This could be a good idea for a _helper function_, something like `setName()` so composition authors don't have to remember the whole annotation.